23092 - Cont in Business Only Alter to Cont in Type

ArwenQin opened the pull request:
*Issue #:* /bcgov/entity#23092

*Description of changes:*
- A C company can alter to a C Type (Cont In ULC, C, CCC, BEN)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

ArwenQin commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-edit-dev--pr-590-mkdch8go.web.app

ArwenQin commented:
Now a Cont in business can alter to other valid Cont in legal types successfully.
Test with:
Cont In ULC to Cont In Benefit or Limited
https://business-edit-dev--pr-590-mkdch8go.web.app/C9900774/alteration?accountid=3040

Cont In Limited to Cont In BEN, ULC, or CCC
https://business-edit-dev--pr-590-mkdch8go.web.app/C9900734/alteration?accountid=3040

severinbeauvais commented on src/components/common/YourCompany/ChangeBusinessType.vue:
```diff
@@ -458,6 +458,16 @@ export default class ChangeBusinessType extends Mixins(CommonMixin) {
+    ) {
+      this.selectedEntityType = convertToContinuedInEntityType(this.selectedEntityType)
+    }
+
```
I'm thinking you can remove `convertToContinuedInEntityType()` ...?

severinbeauvais replied:
See also `convertToRegularEntityType()`.

Ref: https://github.com/bcgov/business-edit-ui/pull/582/files#diff-ca058487aa6730e697018454f9201fad0a0c649b2084211046382fa3e93b6c61

ArwenQin replied:
Are these 2 methods used in the previous ticket to solve the issue that for Cont in, if the user clicks on "Change" for the Business Type then the Business Type Selection is blank?

severinbeauvais replied:
Yes..

severinbeauvais commented:
I cannot test the preview links because Auth API is not returning settings or authorizations.

ArwenQin commented:
> [https://business-edit-dev--pr-590-mkdch8go.web.app/C9900734/](https://business-edit-dev--pr-590-mkdch8go.web.app/C9900734/alteration?accountid=3040)

Oh yes I see. Could you please go to the dashboard of C9900734 and C9900734, and try to alter them?

severinbeauvais commented:
> > I cannot test the preview links because Auth API is not returning settings or authorizations.
> 
> Oh yes I see... I tried, not working either

I will checkout your code and try it locally. (`gh pr checkout 590`)

severinbeauvais commented:
OK, I tested locally the first business (CUL -> CBEN) and the filing JSON looks correct (see below), but I still think there may be an easier solution.

![image](https://github.com/user-attachments/assets/f4973589-5944-49b5-9355-74fb3d1e73c7)

ArwenQin commented:
> OK, I tested locally the first business (C9900774) and the filing JSON looks correct (see below), but I still think there may be an easier solution.
> 
> ![image](https://private-user-images.githubusercontent.com/10002889/369546174-f4973589-5944-49b5-9355-74fb3d1e73c7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY4NjQ0OTEsIm5iZiI6MTcyNjg2NDE5MSwicGF0aCI6Ii8xMDAwMjg4OS8zNjk1NDYxNzQtZjQ5NzM1ODktNTk0NC00OWI1LTkzNTUtNzRmYjNkMWU3M2M3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTIwVDIwMjk1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY5NzFlZWQwOTI5YzRhYjBhYTM3OGZlMmZmZmUyZWY2YzY1ZjlkN2U0YTk2ODc1OWY3YzI5Yzc2NzM0NTE4NmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.EFk3KrXkV7JbqMlfe4rusu4X3KEuGjvu5kFe8yabNNo)

Hi Sev, I just updated the solution. I also corrected the entityTypeOptions property for C, CBEN and CUL

ArwenQin commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-edit-dev--pr-590-mkdch8go.web.app

severinbeauvais commented on src/components/common/YourCompany/ChangeBusinessType.vue:
```diff
@@ -402,7 +391,7 @@ export default class ChangeBusinessType extends Mixins(CommonMixin) {
       return false
     }
-    // Named companies to CC or ULC require a name request.
+    // Named companies to CC/CCC or ULC require a name request.
```
Also CUL I think.

severinbeauvais commented on src/utils/resource-utils.ts:
```diff
@@ -2,7 +2,11 @@ import {
+  AlterationResourceCul,
+  AlterationResourceC,
+  AlterationResourceCben,
+  AlterationResourceCcc
```
Please alphabetize these (and the object below).

severinbeauvais commented on src/components/common/YourCompany/ChangeBusinessType.vue:
```diff
@@ -503,22 +473,26 @@ export default class ChangeBusinessType extends Mixins(CommonMixin) {
   }
 
-  /** Whether current entity selection is a BC CCC. */
+  /** Whether current entity selection is a CCC. */
```
Please be consistent with the comments on lines 474, 480 and 486.

severinbeauvais replied:
(What I'd do is match the comments to the method name.)

severinbeauvais commented:
> Hi Sev, I just updated the solution. I also corrected the entityTypeOptions property for C, CBEN and CUL

OK, this "feels like" the solution I was expecting.

However, I have a done a bit of testing and there are issues (though maybe not caused by this code)...

1. C9900774 is, at the moment, of type "C". But if try to create a name-change NR for it, business search (from Namerequest UI) says it's type CUL. Please check with Vysakh whether this should have been synced.

2. Here is the draft alteration filing (where I changed type from C to CBEN). Do you know if filing.alteration.nameRequest.legalType should also be CBEN? (Maybe ask Vysakh again.)

![image](https://github.com/user-attachments/assets/cc6244dd-5151-4ace-883e-4cd3f2a1eb0b)

Please test thoroughly, including with new NR to change the business name. Thanks.

ArwenQin commented:
/gcbrun

ArwenQin commented:
Thanks Sev. I just updated the comments/order.
I will check with Vysakh

bcregistry-sre commented:
Temporary Url for review: https://business-edit-dev--pr-590-mkdch8go.web.app

ArwenQin commented:
@severinbeauvais 
I just did thorough tests, the type alter works well. Only the Namerequest type pending Vysakh's confirmation.
By the way, I found another bug in the 3. Court Order and Plan of Arrangement section:
![image](https://github.com/user-attachments/assets/8395c4a5-22d8-4d15-bbbf-f1e6c16d43e6)
If I enter the court order number only, and don't check the POA box, there shows no error; but when we file it, there is a schema error.
If we don't enter the court order number, or enter court order number & check the box - all good.
Just to confirm, what is the requirement? Is court order number only + no check box allowed?
Shall I create a new ticket for this?

severinbeauvais commented:
> Just to confirm, what is the requirement? Is court order number only + no check box allowed? Shall I create a new ticket for this?

It should be possible to enter a court number and not check the POA box. This is probably a simple fix you can do in this ticket. According to the [court order schema](https://github.com/bcgov/business-schemas/blob/main/src/registry_schemas/schemas/court_order.json), `effectOfOrder` is a string but isn't nullable, so you probably need to remove that property if it's empty.

Please test a ChangeOfRegistration ("firm alteration") filing for the same bug, and fix that, too.

If you really want, you can create a new ticket for this. (And then just work on it today.)

ArwenQin commented on src/mixins/filing-template-mixin.ts:
```diff
@@ -331,7 +332,7 @@ export default class FilingTemplateMixin extends DateMixin {
         fileNumber: this.getFileNumber,
-        effectOfOrder: this.getHasPlanOfArrangement ? EffectOfOrders.PLAN_OF_ARRANGEMENT : null,
+        effectOfOrder: this.getHasPlanOfArrangement ? EffectOfOrders.PLAN_OF_ARRANGEMENT : undefined,
         hasPlanOfArrangement: this.getHasPlanOfArrangement
```
corrected the EffectOfOrders schema for Court Order section

ArwenQin commented on src/mixins/filing-template-mixin.ts:
```diff
@@ -304,6 +304,7 @@ export default class FilingTemplateMixin extends DateMixin {
         ...this.getNameRequest,
         legalName: this.getNameRequestLegalName || undefined, // don't include if empty
+        legalType: this.getEntityType,
         nrNumber: this.getNameRequestNumber
```
![image](https://github.com/user-attachments/assets/549473e8-cff8-45e4-9482-4af97eabf6f9)
Per Vysakh's notes, I updated the lteration.nameRequest.legalType to be synced with the altered legal type

ArwenQin replied:
But for the first issue, I tested with other businesses, the alterations are not synced to colin either. I'm not familiar with this, may I know the possible reasons?

severinbeauvais replied:
Vysakh identified that the sync didn't work for some reason. I just wanted him to know about it. That's outside the scope of this ticket so don't worry about it.

ArwenQin commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-edit-dev--pr-590-mkdch8go.web.app

severinbeauvais commented on src/mixins/filing-template-mixin.ts:
```diff
@@ -209,7 +209,7 @@ export default class FilingTemplateMixin extends DateMixin {
         filing.correction.courtOrder = {
           fileNumber: this.getFileNumber,
-          effectOfOrder: this.getHasPlanOfArrangement ? EffectOfOrders.PLAN_OF_ARRANGEMENT : null,
+          effectOfOrder: this.getHasPlanOfArrangement ? EffectOfOrders.PLAN_OF_ARRANGEMENT : undefined,
```
Please verify by filing a correction with a court order number and with / without a POA.

ArwenQin replied:
Tested ok with CP1002614

severinbeauvais commented on src/mixins/filing-template-mixin.ts:
```diff
@@ -647,7 +648,7 @@ export default class FilingTemplateMixin extends DateMixin {
       filing.changeOfRegistration.courtOrder = {
         fileNumber: this.getFileNumber,
-        effectOfOrder: this.getHasPlanOfArrangement ? EffectOfOrders.PLAN_OF_ARRANGEMENT : null,
+        effectOfOrder: this.getHasPlanOfArrangement ? EffectOfOrders.PLAN_OF_ARRANGEMENT : undefined,
```
Please verify by filing a change of registration with a court order number and with / without a POA.

ArwenQin replied:
Tested ok with FM1085277

severinbeauvais reviewed (approved):
The code looks good but please ensure you have tested these scenarios before merging (and also add Test Notes in the ticket to let Riyaz know what you've verified locally):
- BC to another type
- BEN to another type
- CC to another type
- ULC to another type
- C to another type
- CBEN to another type
- CCC to another type
- CUL to another type
- use name change NRs for the above wherever possible (if you find bugs related to this then create another ticket)

Thanks for the good work 👍

severinbeauvais commented:
/gcbrun
